File: 0002-AddTwoNumbers/Solution.java
/*  Solution: Digit-by-digit
    Time complexity:    O(n)
    Space complexity:   O(n)

    Note: n is the length of the larger of the two lists.
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode sum = new ListNode();
        ListNode currentNode = sum;
        int op1 = 0;
        int op2 = 0;
        int carry = 0;
        while (true) {
            //  Operand 1
            if (l1 == null) {
                op1 = 0;
            }
            else {
                op1 = l1.val;
                l1 = l1.next;
            }
            
            //  Operand 2
            if (l2 == null) {
                op2 = 0;
            }
            else {
                op2 = l2.val;
                l2 = l2.next;
            }
            
            //  Actual addition
            currentNode.val = op1 + op2 + carry;
            
            //  Determine carry
            if (currentNode.val >= 10) {
                currentNode.val %= 10;
                carry = 1;
            }
            else {
                carry = 0;
            }
            
            //  Break if no more operand nodes left
            if (l1 == null && l2 == null) {
                break;
            }
            
            //  Otherwise move to next digit
            currentNode.next = new ListNode();
            currentNode = currentNode.next;
        }
        
        //  Add extra 1 if carry remains
        if (carry == 1) {
            currentNode.next = new ListNode(1);
        }
        return sum;
    }
}

File: 0002-AddTwoNumbers/Solution.go
/*  Solution: Elementary Math
    Time Complexity:    O(n)
    Space Complexity:   O(n)

    Note: n is the length of the larger of the two lists
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
  l3 := &ListNode{}
  current := l3
  carry := 0
  var op1 int
  var op2 int
  for l1 != nil || l2 != nil {
    if l1 == nil {
      op1 = 0
      op2 = l2.Val
      l2 = l2.Next
    } else if l2 == nil {
      op1 = l1.Val
      op2 = 0
      l1 = l1.Next
    } else {
      op1 = l1.Val
      op2 = l2.Val
      l1 = l1.Next
      l2 = l2.Next
    }
    if carry + op1 + op2 < 10 {
      current.Val = carry + op1 + op2
      carry = 0
    } else {
      current.Val = (carry + op1 + op2) % 10
      carry = 1
    }
    if l1 != nil || l2 != nil {
      current.Next = &ListNode{}
      current = current.Next
    }
  }
  if carry == 1 {
    current.Next = &ListNode{Val: 1}
  }
  return l3
}
